packages/wallet/src/services/TransactionReemitter.ts

```diff
@@ -0,0 +1,81 @@
+import { Cardano } from '@cardano-sdk/core';
+import { DocumentStore } from '../persistence';
+import { Logger, dummyLogger } from 'ts-log';
+import { NewTxAlonzoWithSlot } from './types';
+import { Observable, filter, from, map, merge, mergeMap, scan, tap, withLatestFrom } from 'rxjs';
+
+interface TransactionReemitterProps {
+  rollback$: Observable<Cardano.TxAlonzo>;
+  confirmed$: Observable<NewTxAlonzoWithSlot>;
+  submitting$: Observable<Cardano.NewTxAlonzo>;
+  store: DocumentStore<NewTxAlonzoWithSlot[]>;
+  stabilityWindowSlotsCount?: number;
+  logger?: Logger;
+}
+
+enum txSource {
+  store,
+  confirmed,
+  submitting
+}
+
+// 3k/f (where k is the security parameter in genesis, and f is the active slot co-efficient parameter
+// in genesis that determines the probability for amount of blocks created in an epoch.)
+const kStabilityWindowSlotsCount = 129_600; // 3k/f on current mainnet
+
+export const createTransactionReemitter = ({
+  rollback$,
+  confirmed$,
+  submitting$,
+  store,
+  stabilityWindowSlotsCount = kStabilityWindowSlotsCount,
+  logger = dummyLogger
+}: TransactionReemitterProps): Observable<Cardano.NewTxAlonzo> => {
+  const volatileTransactions$ = merge(
+    store.get().pipe(
+      mergeMap((txs) => from(txs)),
+      map((tx) => ({ source: txSource.store, tx }))
+    ),
+    confirmed$.pipe(map((tx) => ({ source: txSource.confirmed, tx }))),
+    submitting$.pipe(map((tx) => ({ source: txSource.submitting, tx: { ...tx, slot: null! } })))
+  ).pipe(
+    scan((volatiles, { tx, source }) => {
+      switch (source) {
+        case txSource.store: {
+          // Do not calculate stability window for old transactions coming from the store
+          volatiles = [...volatiles, tx];
+          break;
+        }
+        case txSource.submitting: {
+          // Transactions in submitting are the ones reemitted. Remove them from volatiles
+          volatiles = volatiles.filter((v) => v.id !== tx.id);
+          store.set(volatiles);
+          break;
+        }
+        case txSource.confirmed: {
+          const oldestAcceptedSlot = tx.slot > stabilityWindowSlotsCount ? tx.slot - stabilityWindowSlotsCount : 0;
+          // Remove transactions considered stable
+          volatiles = [...volatiles.filter(({ slot }) => slot > oldestAcceptedSlot), tx];
+          store.set(volatiles);
+          break;
+        }
+      }
+      return volatiles;
+    }, [] as NewTxAlonzoWithSlot[])
+  );
+
+  return rollback$.pipe(
+    withLatestFrom(volatileTransactions$),
+    tap(
+      ([{ id }, volatiles]) =>
+        volatiles.some((tx) => tx.id === id) ||
```

Comment:
If you `map` before `tap` then you won't need to iterate the array again

Reply:
> If you `map` before `tap` then you won't need to iterate the array again

I did it like this because after map I lose the `id`, and I wanted to log it.
I could also emit the id from `map`, then strip it in another map. Something like this:
```
map // id -> [id, volatileTx]
tap // log if tx is undefined
map // [id, tx] -> tx

Reply:
Good idea, 2 maps is better than iterating the array twice

Reply:
Done in: [ddb6af8](https://github.com/input-output-hk/cardano-js-sdk/pull/351/commits/ddb6af800a591149f87a2668b4da04d0470a29c7)